### ISSUE TITLE
appStoreVersion and installedVersion are null in Android

### ISSUE THREAD

**Issue by suiebay** [larryaasen/upgrader#101]
Hello, this is my usage of UpgradeAlert widget:

UpgradeAlert(
            appcastConfig: cfg,
            messages: MyUpgraderMessages(),
            debugLogging: true,
            durationToAlertAgain: Duration(days: 1),
            dialogStyle: UpgradeDialogStyle.cupertino,
            showIgnore: false,
            showReleaseNotes: false,
            child: !box.containsKey('show') ? IntroScreen() : MainScreen(),
),

When app runs, debugLogging prints me these informations, and in every hot restart, I get appStoreVersion and installedVersion null, because of it Upgrader Alert does not show. But if I will do hot reload(cmd+s) after launching, appStoreVersion and installedVersion will be found and Alert will be shown. Could you please tell me why versions are not found in first launch? What should I do? Or this is a bug?

I/flutter ( 7734): upgrader: build UpgradeAlert
I/flutter ( 7734): upgrader: languageCode: en
I/flutter ( 7734): upgrader: package info packageName: com.woopapp.woop_client
I/flutter ( 7734): upgrader: package info appName: WOOP
I/flutter ( 7734): upgrader: package info version: 1.0.11
I/flutter ( 7734): upgrader: appcast is available for this platform
I/flutter ( 7734): upgrader: build UpgradeAlert
I/flutter ( 7734): upgrader: blocked: false
I/flutter ( 7734): upgrader: debugDisplayAlways: false
I/flutter ( 7734): upgrader: debugDisplayOnce: false
I/flutter ( 7734): upgrader: hasAlerted: false
I/flutter ( 7734): upgrader: appStoreVersion: null
I/flutter ( 7734): upgrader: installedVersion: null
I/flutter ( 7734): upgrader: minAppVersion: null
I/flutter ( 7734): upgrader: isUpdateAvailable: false
I/flutter ( 7734): upgrader: shouldDisplayUpgrade: false
I/flutter ( 7734): upgrader: shouldDisplayReleaseNotes: false
I/flutter ( 7734): upgrader: appcast item count: 1
I/flutter ( 7734): upgrader: appcast best item version: 1.1.0

**Comment by sisalik1** [comment]
I changed my app version to lower and it shows this, but not the appStoreVersion:
```
I/flutter (30184): upgrader: appStoreVersion: null
I/flutter (30184): upgrader: installedVersion: 1.5.5
```
How the link should look like anyway? Maybe I have a wrong link to check the version...

**Comment by larryaasen** [comment]
Looking at this log, it is strange that the line `upgrader: appcast item count: 1` is not right after `upgrader: appcast is available for this platform`. I think that is the reason that initially appStoreVersion is null. I don't understand why this line doesn't complete in the right order: `await appcast.parseAppcastItemsFromUri(appcastConfig!.url!);`. Maybe it is the logging that is out of order.

**Comment by svyazist2005** [comment]
Do you have any suggestions how to fix it?

> Looking at this log, it is strange that the line `upgrader: appcast item count: 1` is not right after `upgrader: appcast is available for this platform`. I think that is the reason that initially appStoreVersion is null. I don't understand why this line doesn't complete in the right order: `await appcast.parseAppcastItemsFromUri(appcastConfig!.url!);`. Maybe it is the logging that is out of order.

**Comment by Rainburn** [comment]
i got the same problem, here is my log before and after hot reload

I/flutter (32400): null
I/flutter (32400): upgrader: build UpgradeAlert
I/flutter (32400): upgrader: languageCode: en
I/flutter (32400): upgrader: operatingSystem: android sdk_gphone_x86_arm-userdebug 11 RSR1.201013.001 6903271 dev-keys
I/flutter (32400): upgrader: package info packageName: id.vision.app.vision
I/flutter (32400): upgrader: package info appName: Talpro
I/flutter (32400): upgrader: package info version: 1.0.0
I/flutter (32400): upgrader: appcast is available for this platform
I/flutter (32400): upgrader: build UpgradeAlert
I/flutter (32400): upgrader: blocked: false
I/flutter (32400): upgrader: debugDisplayAlways: false
I/flutter (32400): upgrader: debugDisplayOnce: false
I/flutter (32400): upgrader: hasAlerted: false
I/flutter (32400): upgrader: appStoreVersion: null
I/flutter (32400): upgrader: installedVersion: null
I/flutter (32400): upgrader: minAppVersion: null
I/flutter (32400): upgrader: isUpdateAvailable: false
I/flutter (32400): upgrader: shouldDisplayUpgrade: false
I/flutter (32400): upgrader: shouldDisplayReleaseNotes: false
I/flutter (32400): upgrader: appcast item count: 1
I/flutter (32400): upgrader: appcast best item version: 2.0

Performing hot reload...                                               ⢿I/flutter (32400): upgrader: build UpgradeAlert
        
Reloaded 0 libraries in 312ms.
I/flutter (32400): upgrader: blocked: false
I/flutter (32400): upgrader: debugDisplayAlways: false
I/flutter (32400): upgrader: debugDisplayOnce: false
I/flutter (32400): upgrader: hasAlerted: false
I/flutter (32400): upgrader: appStoreVersion: 2.0
I/flutter (32400): upgrader: installedVersion: 1.0.0
I/flutter (32400): upgrader: minAppVersion: null
I/flutter (32400): upgrader: isUpdateAvailable: true
I/flutter (32400): upgrader: shouldDisplayUpgrade: true
I/flutter (32400): upgrader: shouldDisplayReleaseNotes: true
I/flutter (32400): upgrader: showDialog title: Talpro Update
I/flutter (32400): upgrader: showDialog message: Versi Baru Tersedia.
I/flutter (32400): upgrader: showDialog releaseNotes: Perbaikan pada Attendance

**Comment by jpr0ger** [comment]
Hello.

To begin with, thank you for your package, I have no doubt that it made life easier of many people.

However, like some users here, I also faced the problem of not displaying the UpgradeAlert at the first start, but only after Hot reload. As in the messages above, when running `UpgradeAlert` the variables `appStoreVersion`, `installedVersion`, `minAppVersion`: were null, and took normal values only after Hot reload. 

After spending several hours with the debugger, I discovered the following behavior:
The `UpgradeAlert `widget is built using a `FutureBuilder `to which the `future` parameter is passed an initialization function `Upgrader().initialize()` in which, in turn, there are several consecutive asynchronous calls to other internal functions. 
Having looked through most of them, I honestly did not come across anything unusual, but after running this code in the debugger, I found that during the execution of this function an interruption occurs and control ahead of schedule is transferred to `FutureBuilder`, in particular, to check: 
`if (processed.connectionState == ConnectionState.done)`

Typically, in such cases, the value `ConnectionState.waiting` comes and execution continues further, but after calling method `parseAppcastItemsFromUri` specifically the lines `await client!.get(Uri.parse(appCastURL))`
instead of waiting for the result, the execution is passed to the FutureBuilder, and in this case, `ConnectionState.done` already arrives. 
Accordingly, the problem arises in cases when the `Upgrader().checkVersion()` function is called even before `Upgrader().initialize()` is fully executed. Seeing this issue, I understand that maybe I'm not the only one faced with this problem.

I solved the problem for myself as follows:

In `UpgradeAlert` widget I chained these 2 functions: 

```
Future<void> _checkVersion(BuildContext context) async {
  return Upgrader()
      .initialize()
      .then((value) => Upgrader().checkVersion(context: context));
}
```

And passed the given chain of calls to the` future` `FutureBuilder` parameter. 
Removing the check which still did not solve anything anyway:
`if (processed.connectionState == ConnectionState.done)` 

I prepared this solution in the Pull Request to your repository. It is not without flaws in my opinion, but it may be useful to someone else. 
At least it does not affect the operation of the root Upgrader class, but only changes the UpgradeAlert widget.

**Comment by larryaasen** [comment]
I have identified the problem and I am working on a fix. I plan to release an updated version with a fix soon.

**Comment by jpr0ger** [comment]
Excellent. Glad to hear it.

**Comment by larryaasen** [comment]
I fixed the issue and pushed a change on to the master branch here: https://github.com/larryaasen/upgrader/commit/ae1314831ac84c7c456c354be3ec659d48090e32

**Comment by larryaasen** [comment]
I just published [version 3.4.1](https://pub.dev/packages/upgrader/versions/3.4.1) to resolve this issue.
